### DISCUSSION TITLE
how to add Contact header to 3xx response?

### DISCUSSION THREAD

**Discussion by juha-h** [baresip/baresip#2593] [Q&A]
How can I add Contact header to 3xx response before calling `ua_hangup`?

Would function `call_set_custom_hdrs` (that for some reason is missing from `baresip.h`) do it?

**Comment by juha-h** [comment]
I tried by calling `ua_set_custom_hdrs` before `call_set_custom_hdrs`, but the header didn't show up in the response:
```
    struct pl temp_pl = { NULL, 0 };
    struct sip_hdr *hdr;
    hdr = mem_zalloc(sizeof(*hdr), NULL);
    pl_set_str(&temp_pl, "Contact");
    pl_dup(&hdr->name, &temp_pl);
    pl_set_str(&temp_pl, "sip:foo@bar");
    pl_dup(&hdr->val, &temp_pl);
    hdr->id = SIP_HDR_CONTACT;
    struct list hdrs;
    list_init(&hdrs);
    list_append(&hdrs, &hdr->le, hdr);
    ua_set_custom_hdrs((struct ua *)ua, &hdrs);
```

**Reply by cspiel1** [reply]
The custom headers from `struct ua` are copied to `struct call` 
- for the outgoing INVITE, or
- for incoming calls in the `sipsess_conn_handler`.

I guess it would be possible to add a custom header during the call exists. The function `call_set_custom_hdrs` has to be made public. Is it allowed by RFC that custom headers change while a call exists?

edit: Better not to make `call_set_custom_hdrs` public because it overwrites the current custom headers. Maybe `custom_hdrs_add()` can be used. It is already public.

**Reply by juha-h** [reply]
Actually it is not custom header, but Contact header that according to RFC 3261 _needs_ to added to 3xx (e.g. 302) response:
```
[21.3.3](https://www.rfc-editor.org/rfc/rfc3261#section-21.3.3) 302 Moved Temporarily

   The requesting client SHOULD retry the request at the new address(es)
   given by the Contact header field ([Section 20.10](https://www.rfc-editor.org/rfc/rfc3261#section-20.10)).  The Request-URI
   of the new request uses the value of the Contact header field in the
   response.
```

**Comment by juha-h** [comment]
`call_get_custom_hdrs()` returns const to which new headers cannot be added by `custom_hdrs_add()`.  In the above test, I used `ua_set_custom_hdrs()` which ends up calling `custom_hdrs_add()`, but I suspect that this call does not have any effect on the response.

**Comment by juha-h** [comment]
Turns out that `ua_hangup()` calls` call_hangup(),` which in turn calls`sipsess_reject()`.   `sipsess_reject()` has as the last arg `Formatted strings with extra SIP Headers`, but `call_hangup()` passes NULL to it.

So looks like the solution to this is that instead of `ua_hangup()`, the app needs to make all relevant lower level calls that those two functions now call internally.  Or may be a new API function, e.g., `ua_redirect()` could be added that does that?

**Comment by cspiel1** [comment]
Ah, I also started a branch for "redirect".
https://github.com/cspiel1/baresip/tree/moved_temporarily

I could give you more details today. But now, I've to prepare the release.

**Comment by juha-h** [comment]
I made yesterday test implementation of
```
ua_redirect(ua, call, scode, reason, uri) ->
    call_redirect(call, scode, reason, uri) ->
        (void)sipsess_reject(call->sess, scode, reason, "Contact: %s\r\nContent-Length: 0\r\n\r\n", uri)
```
It seems to work OK, but when I tested it I noticed very worrying behavior.  When baresip receiced 302, it did this:
```
call: blind transfer to sip:foo@bar.com
```
This is dangerous.  When 302 is received, the user must be consulted and asked permission to make another call to 302 contact.

**Reply by cspiel1** [reply]
> This is dangerous. When 302 is received, the user must be consulted and asked permission to make another call to 302 contact.

Then maybe the incoming SIP 302 only should be passed via baresip event to the application. Which then may decide on its own if the target should be called or not. It simply could use the API function for starting a call.

**Comment by juha-h** [comment]
My test solution is here https://github.com/baresip/baresip/compare/main...juha-h:baresip:main.

**Reply by cspiel1** [reply]
We would prefer to set an absence note at the UA with a command `uamoved` that can be cleared when the user comes back with a command `uaavail`. Then baresip sends the 302 automatically if an incoming call occurs.

**Comment by cspiel1** [comment]
This is what I have so far: https://github.com/baresip/baresip/pull/2595

Needs tests. Not sure what already works. It was too long lying around.

**Reply by cspiel1** [reply]
I'll try to test it today.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> This is what I have so far:  https://github.com/baresip/baresip/pull/2595

In my opinion my solution is cleaner.  It is simpler and leaves the
decision on how (301, 302, ...) and when (e.g. weekdays and business
hours) to respond to the application.

**Reply by cspiel1** [reply]
okay, then pls create a PR first for call.c/ua.c! After that I maybe put some application logic to module menu with the commands for setting/clearing an absence note.

**Reply by cspiel1** [reply]
Note: Module menu immediately starts a ringtone if an incoming INVITE is received. That's why module menu should have logic to handle user absence information.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Note: Module menu immediately starts a ringtone if an incoming INVITE
> is received. That's why module menu should have logic to handle user
> absence information.

Fine, since I'm not using menu in my app, that part is not included in
the PR.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> okay, then pls create a PR first for call.c/ua.c! After that I maybe
> put some application logic to module menu with the commands for
> setting/clearing an absence note.

Done https://github.com/baresip/baresip/pull/2596.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> > This is dangerous. When 302 is received, the user must be consulted
> > and asked permission to make another call to 302 contact. 
> 
> Then maybe the incoming SIP 302 only should be passed via baresip
> event to the application. Which then may decide on its own if the
> target should be called or not.

Yes.
